### ISSUE TITLE
[Feature request] Windows titlebar icons

### ISSUE THREAD

**Issue by hasufell** [swaywm/sway#4882]
I generally find it hard to distinguish the type of window just based on the title. I'm wasting time reading titles and being confused about whether I'm switching to a console, browser window or something else. Workspaces alone don't fix it, they are only useful to *group* windows, not distinguish them visually. Icons fix this easily with little visual overhead.

There were [several](https://github.com/i3/i3/compare/4.13...mickael9:window-icons.diff) i3 window titlebar icon patches. Also see the [screenshot](https://imgur.com/a/NBB7n).

I wasn't able to find an existing issue report about this.

**Comment by emersion** [comment]
Sorry, this isn't something we're interested in.

**Comment by hasufell** [comment]
> Sorry, this isn't something we're interested in.

Why?

**Comment by emersion** [comment]
We try to keep Sway simple. You're free to add this feature to a fork of Sway if you want it.

**Comment by hasufell** [comment]
> We try to keep Sway simple.

What is simple and what isn't? How is showing a text simpler than showing an icon?

**Comment by emersion** [comment]
Icon lookup is far from simple. Sway would need to implement it, then load the image (supporting multiple file formats), and render it.

**Comment by hasufell** [comment]
Wondering what https://github.com/swaywm/sway/issues/307 was about then

**Comment by vvrein** [comment]
@hasufell you could try something like this:
```
for_window    [app_id="(?i)firefox"]            title_format "<span foreground='#FF6611'><tt></tt></span> %title"
for_window    [app_id="Alacritty"]              title_format "<span foreground='#00FF00'><tt></tt></span> %title"
for_window    [app_id="telegramdesktop"]        title_format "<span foreground='#84b3ff'><tt></tt></span> %title"
```

**Comment by laur89** [comment]
Worth mentioning that upstream i3 [has implemented the feature](https://github.com/Airblader/i3/commit/abbf6a85d712ef1790b02122eb432e4bafe20576).

**Comment by brett-mahar** [comment]
As a workaround, you can name windows incorporating emojis, eg in sway's config:

for_window [title="cmus"] title_format "🔊 cmus "
for_window [title="rss"] title_format "📜 rss "

And your font has to have emoji support, eg:

font pango:Liberation Sans 21

**Comment by fluix-dev** [comment]
> Worth mentioning that upstream i3 [has implemented the feature](https://github.com/Airblader/i3/commit/abbf6a85d712ef1790b02122eb432e4bafe20576).

Though this is a link to Airblader's fork, true upstream did indeed implement this in https://github.com/i3/i3/pull/4439.

**Comment by emersion** [comment]
FWIW, I don't really like the idea of this feature from a technical point of view. It comes at a significant cost: using the XDG icon spec to find the icon from the app-id, then load it (we need to support a variety of file formats), then upload it to the GPU for rendering.

**Comment by emersion** [comment]
Note, i3 doesn't need all of this, because X11 clients upload their icon pixel data themselves to the X11 server as a window property. Wayland doesn't do such things (for good reasons!), but that means more work for Sway.

**Comment by emersion** [comment]
Yet another note, a way out of this would be to have a helper client responsible for rendering the titlebars. This would also allow users to arbitrarily customize the titlebars. This needs a lot more investigation before it can become a viable option, though.

**Comment by fluix-dev** [comment]
> This needs a lot more investigation before it can become a viable option, though.

Very true, but this sounds like it could be really clever!

**Comment by WhyNotHugo** [comment]
> Yet another note, a way out of this would be to have a helper client responsible for rendering the titlebars. This would also allow users to arbitrarily customize the titlebars. As a bonus, this would remove the need for cairo/pango in Sway completely.

I believe that, for an implementation to be efficient, it should need to draw the titlebars into a surface itself (rather than pipe bytes to sway and have sway draw that), right?

What do you think of drawing _all_ decorations on this external helper (titlebars **and** borders). The default helper that sway ships can keep the current, very simple, implementation, but a potential third party one can add shadows and other eye candy without adding maintenance burden on sway.

**Comment by emersion** [comment]
Yeah, that's what I had in mind.

**Comment by twitchyliquid64** [comment]
How do you imagine this working? A new WLR-specific Wayland protocol for signalling what needs to be rendered by the helper program + shuttling the surfaces around, or perhaps something more lightweight (just using wayland surfaces as a convenient render surface)?

I'll be between roles soon, and might be interested in giving this a go.

**Comment by WhyNotHugo** [comment]
Basically the helper needs to render two things: borders, and titlebar. The titlebar in sway doesn't follow the same rules as more traditional (e.g.: xfce-like) desktops. For a window in a tabbed container, the titlebar should actually enumerate all the windows in the tabbed container, so the protocol needs to actually describe the three itself. 

I've seen other x11 window managers that something similar when it comes to tabbing arbitrary windows (e.g.: pekwm), so the wayland protocol may be reusable by other compositors (even if the helpers themselves are not necessarily all portable). 

---

Re regarding the size of the borders "surface". Assume a window has a 200x200px space. If the helper renders a 2px border, should sway resize the inner window to 196x196px? Does the helper render 200x200px with a transparent/ignored area in the middle? Or does it render four different surfaces (top, right, bottom, left)? I suspect the latter is more efficient.

The helper could indicate a global "oh, my borders are all top:2px, bottom: 2px, left: 2px, right:2px", so sway knows how much it'll need to pad  windows.. I think per-window border sizes would be far more problematic.

Side question: Can borders for floating windows have alpha? I'm sure someone will eventually want to use this for shadows and such.

**Comment by twitchyliquid64** [comment]
For a first pass, would we really want to delegate _all_ logic for configuring border sizes etc to the helper? could we keep that in Sway's configuration, and simply pass that global info (as well as the window information and the surfaces) to the helper for rendering?

**Comment by twitchyliquid64** [comment]
More generally, what should be the behavior in the case that the helper is 'slow' in rendering the surface and yeeting it back? My guess is that somehow, sway should avoid blocking frames and instead just paint the border/titlebar with the correct color as defined by the config?

**Comment by mid-kid** [comment]
> Note, i3 doesn't need all of this, because X11 clients upload their icon pixel data themselves to the X11 server as a window property. Wayland doesn't do such things (for good reasons!), but that means more work for Sway.

What good reasons? Applications being able to dynamically update their window icon alongside their title has been a thing since the dawn of time, I'm not sure what exactly is the issue with communicating that along with the protocol.

**Comment by PAI5REECHO** [comment]
I'm surprised that people are able to use Sway quickly and efficiently without window titlebar icons. Having to visually scan dozens of window titles you have open wastes a lot of time and reduces productivity.

**Comment by emersion** [comment]
Discussion about this on IRC:

```
2022-03-14 21:33:37	twitchyliquid64	Hi! I'm interested in working on https://github.com/swaywm/sway/issues/4882#issuecomment-886101225 and wanted to kick off the conversation in a high-bandwidth channel like here if thats okay
2022-03-14 21:43:46	emersion	sure thing
2022-03-14 21:43:53	emersion	are you familiar with wayland protocols?
2022-03-14 21:59:28	<--	illiliti (~illiliti@user/illiliti) has quit (Quit: leaving)
2022-03-14 22:09:49	twitchyliquid64	A little!
2022-03-14 22:10:47	twitchyliquid64	Were you thinking of a wayland protocol extension to communicate the titlebars + surfaces back and forth, somewhat in the same spirit of how wlr layer shell works?
2022-03-14 22:27:06	emersion	yeah i was thinking about a protocol to set the surface for each side
2022-03-14 22:27:58	emersion	one really important thing is synchronization
2022-03-14 22:28:14	emersion	also disclaimer: i'm not 100% sure this is a good way forward
2022-03-14 22:28:39	emersion	ifreund, kennylevinsen, vyivel: thoughts on the approach described in the link above?
2022-03-14 22:29:29	emersion	the possible alternatives would be:
2022-03-14 22:29:55	emersion	(1) just… load the PNG/JPG/SVG files inside sway and render the icons there
2022-03-14 22:30:30	emersion	(2) add a protocol for a helper client which just renders the icon, not the whole decorations
2022-03-14 22:31:13	emersion	(3) restrict how decorations can look like so that we can re-render them in sway without asking the helper client to re-render
2022-03-14 22:31:35	emersion	(e.g. surfaces for each corner + repeating surfaces for the edges)
2022-03-14 22:31:49	twitchyliquid64	Is communicating with the helper client over stdin + shared memory (as opposed to a brand new wayland protocol, which would need code in wlr) a non-starter?
2022-03-14 22:31:57	emersion	(+ one surface anchored on the top-left?)
2022-03-14 22:32:16	emersion	such a protocol would just reinvent wayland
2022-03-14 22:32:30	emersion	wayland is basically a pipe + shared memory
2022-03-14 22:32:36	emersion	s/pipe/socketpair/
2022-03-14 22:32:51	twitchyliquid64	Yeah fair enough
2022-03-14 22:33:12	twitchyliquid64	I'll need to dig in and learn wlr and a bunch more is all, but if its the right way to do it its the right way to do it.
2022-03-14 22:33:17	emersion	but if we just reinvent our thing, we need to also reinvent a marshalling, synchronization primitives, what if clients want to use the GPU, etc
2022-03-14 22:33:42	emersion	submitting a wlr patch shouldn't be more scary than submitting a sway patch
2022-03-14 22:33:54	twitchyliquid64	okies
2022-03-14 22:34:00	twitchyliquid64	As for the alternatives, Im not a huge fan of (1) or (2) because a lot of the appeal of this (for me at least) is flexibility. As for (3) I don't know enough about these nuances yet.
2022-03-14 22:34:09	emersion	it's basically the same, except in wlr it can be re-used by other compositors, if there is interest
2022-03-14 22:34:32	emersion	and we can provide guidance ;)
2022-03-14 22:34:37	emersion	yeah
2022-03-14 22:34:45	emersion	i'm not a huge fan of (1) for sure
2022-03-14 22:35:01	emersion	i don't want to have an SVG parser in something as critical as a compositor
2022-03-14 22:35:43	emersion	and yeah, the flexibility is reduced greatly with (1) and (2), but it's also a simpler problem to solve :P
2022-03-14 22:37:32	twitchyliquid64	As a starting point, could we aim to start with just titlebars (ie: `container->title_focused` and friends) rather than the full set of decorations? Or is it easier just to lump it all into one surface rather than borders and titlebar and marks all separately
2022-03-14 22:38:53	emersion	if you want to start experimenting right away, it might be a good start (especially if you're new to all these things)
2022-03-14 22:39:31	emersion	have you read the wayland book?
2022-03-14 22:40:42	twitchyliquid64	A while back (im familiar with the concepts) but its probably time for another read
2022-03-14 22:43:55	vyivel	emersion: not sure if i like the idea of separate client being responsible for decorations tbh
2022-03-14 22:44:11	vyivel	a library to help with managing this kind of stuff, maybe
2022-03-14 22:44:49	vyivel	wouldn't put in wlroots
2022-03-14 22:47:46	twitchyliquid64	Hmm, how else would you implement it (flexibility in rendering titlebars and the like) short of forking sway?
2022-03-14 22:49:22	ifreund	emersion: I'd lean towards 2) myself. It keeps the code parsing stupidly the complex svg file format outside of the compositor process but doesn't introduce the IMO unnecessary complexity of rending all parts of the decorations in the client
2022-03-14 22:49:48	vyivel	twitchyliquid64: perhaps with something like libdecoration but on the server side
2022-03-14 22:50:22	ifreund	if all parts of the decorations were rendered in the client I think it'd also result in lots of synchronous round trips to get new decorations on focus/urgency/etc
2022-03-14 22:51:02	vyivel	that is, if we're discussing flexible decorations; for just an icon, a sway/wlroots-specific helper client is totally fine by me
2022-03-14 22:58:57	emersion	vyivel: we don't really want to do stuff as complex as rendering fancy titlebars in the compositor though
2022-03-14 22:59:48	vyivel	why not?
2022-03-14 23:00:10	emersion	if the compositor goes down, the whole session is toast
2022-03-14 23:00:15	emersion	thus the compositor needs to be robust
2022-03-14 23:01:13	emersion	robust = doesn't do complex operations (which will take time, slowing down everything else), and doesn't parse heaps of untrusted data with complicated logic (PNG/JPG/SVG)
2022-03-14 23:01:58	emersion	exhibit 1: GNOME does all of this in the main compositor thread, and more (JS engine for their extensions)
2022-03-14 23:02:12	bl4ckb0ne	any compositor maintainer here using the device field of the wlr_event_switch_toggle https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/104060fec57bec2a282c52f7e849f40fd1b1b45e/include/wlr/types/wlr_switch.h#L41
2022-03-14 23:02:18	vyivel	gnome shouldn't be used as an example of good software
2022-03-14 23:02:20	emersion	it's a bad design decision but they can't easily revert it
2022-03-14 23:03:08	kennylevinsen	Hmm drawing decorations in a separate client sounds cool but also like true synchronization hell
2022-03-14 23:03:13	vyivel	^
2022-03-14 23:03:24	emersion	yeah, the synchronization part of it is pretty scary
2022-03-14 23:03:35	vyivel	if not sync i'd be fine with it
2022-03-14 23:04:31	vyivel	maybe could slap xdg-foreign in this somehow, if it wasn't so limited
2022-03-14 23:07:29	kennylevinsen	A client could send us a raster of a titlebar segment with alpha that we blend over the titlebar (and also contains the text, removing pango), but I am not sure if it is worth the complexity
2022-03-14 23:07:59	kennylevinsen	Otherwise we'd definitely be relying on transactions heavier to synchronize client changes to the decorator
2022-03-14 23:09:33	vyivel	just run every client in a nested compositor which draws decorations as it wants ezpz
2022-03-14 23:10:26	 *	kennylevinsen pulls vyivel's power cable out
2022-03-14 23:11:32	vyivel	i run on battery :p
2022-03-14 23:12:14	kennylevinsen	don't run on the battery, it's fragile
2022-03-14 23:12:28	vyivel	-_-
2022-03-14 23:12:32	bl4ckb0ne	all in favour of yeeting the input device from wlr_switch_toggle_event?
2022-03-14 23:12:43	 *	twitchyliquid64 yeets in australian]
2022-03-14 23:13:18	vyivel	people who choose to support switch devices probably have code complex enough to store the device somewhere else anyway
2022-03-14 23:16:50	twitchyliquid64	On the topic of offloading raster to separate clients, is there a reason why the sway bar has its own funky protocol as opposed to being a wayland extension + client?
2022-03-14 23:17:18	vyivel	funky protocol?
2022-03-14 23:18:25	vyivel	afaik it's just layer-shell + sway ipc for workspaces
2022-03-14 23:21:17	twitchyliquid64	vyivel: The JSON protocol thingy
2022-03-14 23:21:18	twitchyliquid64	https://github.com/swaywm/sway/blob/master/swaybar/swaybar-protocol.7.scd
2022-03-14 23:22:01	kennylevinsen	Ah that's an i3 thing
2022-03-14 23:22:22	kennylevinsen	Or wait is it
2022-03-14 23:22:34	vyivel	yep
2022-03-14 23:22:34	vyivel	https://i3wm.org/docs/i3bar-protocol.html
2022-03-14 23:23:37	emersion	it's for communication between swaybar and the status_command
2022-03-14 23:23:42	emersion	not between sway and swaybar
2022-03-14 23:23:52	emersion	it basically carries the bar text
2022-03-14 23:23:59	emersion	which swaybar will render
2022-03-14 23:27:18	twitchyliquid64	Ahhhh gotcha
2022-03-14 23:27:56	twitchyliquid64	Okay heres a cursed idea
2022-03-14 23:28:19	twitchyliquid64	what if sway supported loading a shared library which implemented decorations or raster of titlebars and the like
2022-03-14 23:28:44	vyivel	as in, finding and loading the library in runtime?
2022-03-14 23:29:02	twitchyliquid64	yeah dl_open, and then calling some raster_titlebar()
2022-03-14 23:29:16	twitchyliquid64	the shared library to use can be configured in the sway config or soemthing like that
2022-03-14 23:29:39	vyivel	ehhhh
2022-03-14 23:29:43	emersion	dlopen hell :S
2022-03-14 23:29:47	vyivel	this sounds even less robust
2022-03-14 23:29:49	kennylevinsen	if it's just a transparent chunk we can overlay with the text and image then it could be an executable writing it to stdout that we call
2022-03-14 23:30:05	kennylevinsen	Once when we find it appropriate and cache the output
2022-03-14 23:30:28	vyivel	what if the executable prints out 1 byte per hour
2022-03-14 23:30:34	vyivel	will it slow down the compositor?
2022-03-14 23:31:03	kennylevinsen	in that case I think you deserve the resulting sluggishness for intentionally adding that sleep
2022-03-14 23:31:12	<--	twitchyliquid64 (~twitchyli@c-24-6-228-142.hsd1.ca.comcast.net) has quit (Quit: Client closed)
2022-03-14 23:31:15	kennylevinsen	Or alternatively you really need a new computer
2022-03-14 23:31:28	-->	twitchyliquid64 ([https://web.libera.chat] twitchyliquid64) (~twitchyli@c-24-6-228-142.hsd1.ca.comcast.net) has joined #sway-devel
2022-03-14 23:31:43	emersion	kennylevinsen: exec'ing one executable per toplevel doesn't seem very fun
2022-03-14 23:31:48	kennylevinsen	it could also send an fd we can import or something...
2022-03-14 23:32:02	kennylevinsen	emersion: but only once/on title change though
2022-03-14 23:32:27	vyivel	feeling like we're two steps away from reinventing x11's "server + wm" model
2022-03-14 23:32:49	twitchyliquid64	vyivel: emersion: I think for dl_open if people use that feature they accept the resulting bugs they introduce?
2022-03-14 23:33:35	twitchyliquid64	Yeah Im in camp 'one offload process per sway instance' rather than fork/exec every time we need a raster. But agree everytime we talk about doing it via stdin it sounds like we reinvent wayland all over again
2022-03-14 23:33:36	kennylevinsen	the idea here was not to have it as an optional feature, it would be the only way title bars are made so we can remove Cairo/pango from sway
2022-03-14 23:34:35	emersion	i don't care that much about removing pango/cairo, it would've been a nice side-effect but it's not the end of the world to have these
2022-03-14 23:35:00	emersion	but yeah ideally there is only one way the titlebars get generated
2022-03-14 23:35:27	emersion	and we don't duplicate the code between "fancy" and "baseline" titlebars
2022-03-14 23:37:54	emersion	in any case, the (ugh) dlopen idea wouldn't buy much in terms of synchronization
2022-03-14 23:38:31	emersion	the process of rendering the titlebars would need to be async because it could take a while and we don't want to block the whole compositor
2022-03-14 23:38:41	emersion	and then we end up with the same issues as a helper wayland client
2022-03-14 23:39:08	twitchyliquid64	What is it currently? Synchronous no?
2022-03-14 23:39:20	emersion	yes, but the titlebars are guaranteed to be very simple
2022-03-14 23:39:27	emersion	solid color + one texture
2022-03-14 23:42:12	twitchyliquid64	I wonder how terrible it would be if we kept the current sway implementation (solid color + texture) in the hot path, but let the 'decorations client' push a new texture or something at any time?
2022-03-14 23:42:35	twitchyliquid64	Could look jittery like the swaybar on startup tho
2022-03-14 23:42:37	emersion	it would look pretty bad when resizing
2022-03-14 23:42:55	emersion	"every frame is perfect with wayland" ™
2022-03-14 23:43:20	twitchyliquid64	whelp
2022-03-14 23:43:25	twitchyliquid64	this is a hard-ass problem
```

**Comment by WhyNotHugo** [comment]
> The core of the issue is synchronization: Wayland expects frame perfection, but if you delegate something as core as titlebars to another process, you need to synchronize. And what if the delegated process is too slow to yeet back rasterized surfaces before the next scan-out? then you either need to block the compositor, or deliver imperfect frames. Both of these are undesirable.

Sounds like this delivers perfect frames, at a price of never allowing out-of-compositor decorations.

> The current implementation works because of its simplicity. Rendering a solid color + some pango text is super simple, so it can stay in the hot path and then we never need to deliver imperfect frames, and the blocking of the compositor from this has never been an issue.

Removing pango and fontconfig dependencies can also reduce complexity for sway.

> Simultaneously, if we were going to go ahead and do this, the right place for it would be a new wayland protocol, because inventing anything else would mostly be re-inventing a lot of the concepts from Wayland into a new protocol.´

I agree.

---

How do other DEs delegate decorations without compromising on perfect frames?

**Comment by WhyNotHugo** [comment]
Also, could some wayland-protocol proxy do decorations and synchronisation on its own? That is, by exposing a different wayland socket to clients.

**Comment by emersion** [comment]
> How do other DEs delegate decorations without compromising on perfect frames?

They mostly don't.

- GNOME doesn't support server-side decorations. Not sure what they do for X11 clients, perhaps the same as Weston.
- KWin has in-house decorations only.
- Weston doesn't support server-side decorations for Wayland clients, it draws decorations only for X11 clients via the X11 protocol (ie, by asking the X11 server to render them).

> Also, could some wayland-protocol proxy do decorations and synchronisation on its own? That is, by exposing a different wayland socket to clients.

Would be very complicated.

**Comment by emersion** [comment]
>      Sorry, but can you elaborate on this a little? I haven't read the Wayland book, nor do I have any knowledge about Wayland's protocol design, but this seems like a rather odd/regressive design decision.

No, this isn't the place to discuss this.

> I understand that synchronization is required but why would this need to block the compositor?

Synchronization doesn't mean blocking the compositor. It means that the updates of the client surface and the decorations are synchronized, and happen atomically, to ensure there are no glitchy frames displayed to the user. The main concern is that this adds substantial complexity.

**Comment by adrusi** [comment]
What are people's thoughts on designing the protocol in such a way that for simple window decorations clients like what sway would presumably ship by default, the consequences of a failure of synchronization are minor (either perfect or minimally imperfect frames), and users who choose to use fancier window decorations may end up with more noticeably imperfect frames if their decorations client can't keep up?

As I understand, the state which would need to be synchronized between the compositor and the decorations client is:

- window titles, which window is active, which window is frontmost in each container, which windows have alerts, possibly everything returned by `swaymsg -t get_tree`
- dimensions of the decorations surfaces, whatever those end up being defined as
- possibly user interaction with decorations elements

The maximally flexible protocol design would allow the decorations client to negotiate, for any container, how much space it gets to draw on any side. This approach is great for my personal agenda, because I'd like to co-opt the stacked layout to implement vertical tabs in the style of firefox's tree-style-tabs extension. However —

An alternative would be to give the compositor more control over which decorations surfaces get to exist. Currently, as I understand (and I might misunderstand), you can categorize containers into two classes based on what decorations they have:

- containers with no decorations (like an application window inside of a tabbed or stacked container)
- containers with optional borders on the left, right and bottom, with 0 or more window titles arranged in some manner on the top side.

Here is an outline of a minimum viable protocol that would minimize the imperfections caused by failures of synchronization:

The compositor gets to decide which decorations surfaces exist. Each container that gets decorations gets three surfaces: left, right and bottom borders. Additionally, each _application window_ gets one decorations surface, namely its title. In order to accommodate resizing of containers without rendering obviously bad frames in the case of lost synchronization, each of these surfaces is larger than it needs to be. Window-title and bottom-border surfaces are all as wide as they might ever possibly be — maybe as wide as the widest output, maybe as wide as the widest row of pixels in the output arrangement. Left and right borders are similarly maximally tall. The compositor notifies the decorations client of the actual rendered regions of these surfaces at any given time, but the client is expected to draw _something_ useful in the rest of the surface just in case synchronization is lost. The compositor never blocks on decorations redraws, instead always rendering the decorations surface clipped to the needed dimensions.

**Advantages**

- The default decorations are solid colors for borders with left-aligned text on a solid background for window titles. The default decorations client could feasibly ignore all information about the actual rendered size of each surface and let the compositor clip them as needed.
- Various oft-requested features can still be accommodated: icons in titles, rounded borders, etc. Although decorations clients providing e.g. rounded borders would need to be good at maintaining synchronization to avoid obviously imperfect frames when a container is resized.

**Disadvantages**

- Memory usage.
- Not maximally flexible, couldn't accommodate e.g. my wishes for vertical tabs on the left side of the container without additional protocol complexity.

**Comment by WhyNotHugo** [comment]
So it seems that synchronisation issues are a blocker for delegating all decorations to a dedicated/privileged client.

The original issue here is just about showing **icons** in titlebars. I'd like to re-focus discussion on that.

I'm thinking on an approach as follows:

- Add a new protocol. Let's call this protocol `ext-icon-manager` for now. It implements:
  - A `set_icon(toplevel: zwlr_foreign_toplevel_handle_v1, icon: wl_surface)` request.
  - A `icons_size(size: uint)` event, which indicates which size all icons should have.
- Have a privileged client use `wlr-foreign-toplevel-management-unstable` to enumerate all toplevels.
- That same priviledged client uses the above described `set_icon` request to set icons for known applications. _How_ it determines which icons to use is implementation specific (note that it'll have access to `title` and `app_id`).
- `wlr-foreign-toplevel-management` already enabled a privileged client to keep track of all existing clients, so no need to do anything there.

On the sway side:

- This just requires rendering on more surface, no parsing PNG or SVG or anything.
- Render the icon on the titlebar if we have one, don't render the icon if we don't.

This protocol is perfectly usable by other compositors too. FWIW, a well-written icon-manager-helper could also be cross-compositor.

Some points worth discussing:

- Is it safe to assume that all clients will have icons with the same size? Or is the protocol being artificially limited for some other scenario here?
- This approach expects the privileged client to proactively send icons for all toplevels. Is it maybe better for the compositor to request them explicitly instead? Requesting them with an event could allow the compositor to specify the size for each one, but I'm not sure that's flexibility that we _need_.

**Comment by mildred** [comment]
Just for the record, color fonts now can embed SVG documents which in itself can contain bipmap image data...

It seems that drawing fonts is (or soon might be) more complex than drawing simple images... https://pixelambacht.nl/2014/multicolor-fonts/

**Comment by mildred** [comment]
This makes me think that perhaps implementing icons on the titlebars could be done by an external process that would transform the icon from the window into a font and return to sway the font to use as well as the glyph to use within the font, and this could be done.

However, it seems sway has to be restarted completely (not just reloaded) to get it to accept a new font. There is probably a call somewhere to make to refresh the font list.

**Comment by mildred** [comment]
> What are people's thoughts on designing the protocol in such a way that for simple window decorations clients like what sway would presumably ship by default, the consequences of a failure of synchronization are minor (either perfect or minimally imperfect frames), and users who choose to use fancier window decorations may end up with more noticeably imperfect frames if their decorations client can't keep up?
> https://github.com/swaywm/sway/issues/4882#issuecomment-1100921783

Seems like a good idea, however I don't think we need to go to the length of drawing all the borders and corners. I'm perfectly fine having this fixed in sway or configurable by simple options. What we might need instead is to draw the content of the titlebar itself.

The simpler version of the protocol could be:

- The compositor gets to decide which decorations surfaces exist and their sizes.
- each application window gets one decorations surface, namely its title, excluding the borders drawn by the compositor
- In order to accommodate resizing of containers without rendering obviously bad frames in the case of lost synchronization, last column of pixels in the surface should be repeatable so it can extend (obviously the text would wait the synchronization to show up

This keep things simple. The only drawback compared to the current state is that when the titlebar is extended, there might be a loss of synchronization that would delay showing up the rest of the text, but it should not be that noticeable and if we want to separate the text rendering to a separate process, this is something can will necessarily happen (unless we take the decision to render all of the text even if not shown). Shrinking the titlebar would not cause any visual imperfection as the title would be clipped as of today.

If we want to extend the protocol to draw the borders, the process that draws the decoration can draw a larger surface that could be cut in four (vertically and horizontally) and in place of the cuts, the compositor would extend the pixels. I do not really imagine a use case where the border could not be drawn like this. But then with drawing borders, we will want to also tell which parts are clickable and which parts should trigger a window resize, ...

**Comment by iguanajuice** [comment]
> @hasufell you could try something like this:
> 
> ```
> for_window    [app_id="(?i)firefox"]            title_format "<span foreground='#FF6611'><tt></tt></span> %title"
> for_window    [app_id="Alacritty"]              title_format "<span foreground='#00FF00'><tt></tt></span> %title"
> for_window    [app_id="telegramdesktop"]        title_format "<span foreground='#84b3ff'><tt></tt></span> %title"
> ```

I like the idea of this; but manually setting each icon is *very* tedious.

So I decided to make this repo: https://github.com/iguanajuice/sway-font-awesomee

Right now, it only adds icons to apps that I personally use. Please open issues for apps that I haven't added an icon for yet (which is a lot of them), if you're interested.

**Comment by sripwoud** [comment]
https://github.com/lyr-7d1h/swayest_workstyle

**Comment by tmccombs** [comment]
> >  Sorry, but can you elaborate on this a little? I haven't read the Wayland book, nor do I have any knowledge about Wayland's protocol design, but this seems like a rather odd/regressive design decision.
>
> No, this isn't the place to discuss this.


Why not? My first thought on how to implement icons in the titlebar would be to have a protocol to allow the client to give the compositor a surface to use for the icon, similar to setting the cursor shape. And allow the application to update the icon, so it can use it to indicate status.

The problem with that is it isn't very useful unless most applications implement it. And since gnome only supports client side decorations it is unlikely gtk would support such a protocol.

With the proposal for a separate decorations app, I'm not sure how it would know what icon to use. Does determine it based on the app id? If so, that means an app can't dynamically change the icon, or use different icons for different windows .  Both of which I think are possible in X. How does KDE determine the icon to use?

**Comment by boomshroom** [comment]
[The `xdg-toplevel-icon` protocol has been merged into `staging`.](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/tree/main/staging/xdg-toplevel-icon)

**Comment by alexsch01** [comment]
wlroots 0.19.0 rc1 implements xdg-toplevel-icon-v1

**Comment by kaitlynia** [comment]
What's the stance on this in 2025? Still against something that has been a part of i3 since 2021, even though wlroots implemented the protocol for it? I imagine it's daunting for someone to attempt this PR just reading through the issue now and seeing the mixed response to it

**Comment by stevenxxiu** [comment]
Ideally this would use the program's XDG icon by default, and also allow the image be overridden per `app_id`, window title, like in <https://github.com/allie-wake-up/swaycons>. This way the terminal that has say *Neovim* open, can have the *Neovim* icon instead.

**Comment by darkralts** [comment]
> What's the stance on this in 2025? Still against something that has been a part of i3 since 2021, even though wlroots implemented the protocol for it? I imagine it's daunting for someone to attempt this PR just reading through the issue now and seeing the mixed response to it

i would like icons. i think its a good idea

**Comment by WhyNotHugo** [comment]
With xdg-toplevel-icon-v1, icon lookup is no longer an issue, since clients provide a buffer with their icon.

Most of the initial objections to this feature were due to complexity in icon file look-up, decoding, etc. With the icons provided via xdg-toplevel-icon-v1, none of that is an issue.

**Comment by WhyNotHugo** [comment]
wlroots supports xdg-toplevel-icon-v1, but sway does not. Before we can consider implementing this, we need xdg-toplevel-icon-v1 support in sway.

**Comment by emersion** [comment]
> With xdg-toplevel-icon-v1, icon lookup is no longer an issue, since clients provide a buffer with their icon.

Most apps don't need per-toplevel icons, thus most clients won't provide a buffer via xdg-toplevel-icon-v1. Additionally, pixel buffers via xdg-toplevel-icon-v1 are pretty restricted and an icon name is preferred.

**Comment by WhyNotHugo** [comment]
> Most apps don't need per-toplevel icons, thus most clients won't provide a buffer via xdg-toplevel-icon-v1.

I don't see the causality between these two statements.

The only mechanism to provide icons is via pre-toplevel icons. Even if an application uses the same icon for all toplevels, this is the only mechanism via which it can specify them.

> Additionally, pixel buffers via xdg-toplevel-icon-v1 are pretty restricted and

In what way are they restricted?

> an icon name is preferred.

I don't see this preference being laid out in the specification.

**Comment by emersion** [comment]
> The only mechanism to provide icons is via pre-toplevel icons. Even if an application uses the same icon for all toplevels, this is the only mechanism via which it can specify them.

I meant that most apps will continue to use the existing per-app icons via app ID, without using the protocol. And this is a good thing.

> In what way are they restricted?

There are known issues around HiDPI, and only a single pixel format is supported.

Anyways, only supporting icons set via a buffer via xdg-toplevel-icon-v1 is a no-go for me.

**Comment by stevenxxiu** [comment]
One implementation I've worked with is <https://github.com/albertlauncher/albert/blob/f5681d32055fe1438a1d83f29f90d677d8777d35/src/platform/xdg/iconlookup.cpp>. It's reasonably short and works well enough. I don't think implementing the same thing in *C* will use much more code.

**Comment by mrusme** [comment]
I would like to add to the discussion that the incoherent UI language of Linux app icons in general makes this feature something that, if it should be implemented, I personally would prefer it to be optional. The visual noise that browser tab icons introduce is already bad enough, I wouldn't want to have to deal with more _random_ shapes and colors spread across my desktop.

Remember: We are not talking about _swaycons_, that are _flat_ shapes with a single color, we're talking _pixelmush_ most of the time. Unless you're solely using Linux programs from the front page of flathub, you will likely see 16x16px-sized versions of icons like [this](https://dl.flathub.org/media/org/kde/kwave/8d66aa2e53c6b3e099a8bff72b4fb47c/icons/128x128/org.kde.kwave.png), [this](https://dl.flathub.org/media/com/nickgirga/webready/94bc352da79101e63a425393f00618a5/icons/128x128/com.nickgirga.webready.png), [this](https://dl.flathub.org/media/fr/handbrake/ghb/3900fb0cbd12d758dc2597a9fcab5043/icons/128x128/fr.handbrake.ghb.png), [this](https://dl.flathub.org/media/io/github/muse_sequencer.Muse/13b63bc06c8b73817e69d7e8681cf3b9/icons/128x128/io.github.muse_sequencer.Muse.png), or [this](https://dl.flathub.org/media/com/behringer/XAirEdit/ae8f33db1bb473961cc67ae58b82d81f/icons/128x128/com.behringer.XAirEdit.png).

Even with a custom icon theme that could override programs' default icons, I remain doubtful that a lot of people would actually jump through the hoops of _ricing_ their desktop to achieve a clean and visually pleasing look. 

If anything, I would hence argue that it would make more sense to natively integrate the feature that _swaycons_ offers (as an optional setting), rather than going with the actual program icons. Given that it's using a (nerd-)font, my understanding from the discussion above is that it is more likely to result in a cleaner implementation than the use of actual icons/pixmaps.

If anyone in this threat is designwise motivated enough I would encourage to create a mockup of a _typical_ Sway desktop with little app icons (preferably not the most streamlined ones) in the title bars and post it for people to get an idea of what the end result will look like. My guess is that the mockup would relatively quickly dampen the enthusiasm for this feature.

//edit

Some people here have pointed out that i3 seemingly supports displaying app icons. To try to validate the assumption of pixel icons making for a less visually appealing appearance we can head over to [_r/unixporn_, filter by _[i3]_ and sort by _top_ rated of _all_ time](https://redlib.privacyredirect.com/r/unixporn/search?q=%5Bi3%5D&restrict_sr=on&sort=top&t=all). Interestingly, though, none of the results on the first few pages depict any titlebar icons whatsoever.

**Comment by mid-kid** [comment]
I don't think anyone is asking for this to be mandatory. You can always disable this, and I don't see why swaycons couldn't be a proper icon theme, at least for the appid-based icons, rather than a difficult to edit font. Using fonts for this purpose has always felt like a massive hack.

**Comment by mrusme** [comment]
>  I don't see why swaycons couldn't be a proper icon theme

I assume because it would be limited to a single color, unless there is support for colorizing pixmaps. What _swaycons_ is doing wouldn't be replicable with an icon theme, unless you offer SVG icons in every possible color or a way to colorize them.

**Comment by mid-kid** [comment]
Also, unixporn doesn't feel like a good representation of what people actually use. I'm still on i3 and on-and-off maintained the iconpatch before it got merged, because they make tabs more visible to me, and I don't mind having the constant monochrome be interrupted by a splash of color. I very much prefer an "incoherent" icon over none at all.

If you need a screenshot, this is what my tab bar looks like right now (shame about ZapZap not setting an icon):
<img width="2007" height="45" alt="Image" src="https://github.com/user-attachments/assets/274686ea-ef5c-437e-a77c-ab7b1d803636" />
And I have some more screenshots with them [here](https://github.com/mid-kid/RocketISO/blob/master/docs/RELEASE-2025.md#screenshots).

> unless you offer SVG icons in every possible color or a way to colorize them.

This is a solved problem, it's just that very few things use it properly. The GTK3 app icons on my status bar recolor themselves properly, according to the color defined by polybar, by changing the CSS of [an icon from the breeze theme](https://github.com/KDE/breeze-icons/blob/master/icons/status/22/audio-volume-high.svg?short_path=df5a87e). I haven't gotten this to work with other toolkits, but I'm sure this is something that could be supported in the SVG renderer path (or you can override it in ~/.icons with your custom color)

**Comment by mrusme** [comment]
@mid-kid thank you for providing the screenshot! 

The reason I mentioned a _typical_ Sway desktop is because I would be curious to see something that is (as you correctly pointed out wrt the linked subreddit) _a good representation of what people actually use_. Personally, I split my screen vertically into two main areas, for which I use the tabbed mode. At times, I have first-level containers with multiple (tabbed) second-level containers, with (tabbed) third-level windows. Every once in a while there are some floating windows on top of all of this. Long story short, imagining this with various window (app) icons doesn't sound particularly _clean_ and easy on the eyes to me.

Besides, although it's just a visual nitpick, the icons displayed in your screenshot are too big for the titlebar. These minor details are something that also need to be considered: What's the ratio of icon vs text size? Would this feature have a way to specify icon size? Would the icons get bigger if the titlebar gets taller? What about the left and right margins, especially the space between the icons and the text?

I think in order to make this feature a pleasant experience there would need to go a lot more thought into it. I do not like the way i3 seemingly implemented it, judging purely from the screenshot(s) you provided, but that is subjective.

**Comment by mid-kid** [comment]
> Besides, although it's just a visual nitpick, the icons displayed in your screenshot are too big for the titlebar. 

As far as I can tell, there's an option to adjust the horizontal padding, but not the vertical padding. The title bar is simply as big as the font size requires, and the icon fills it up vertically. Feels like a missing feature, considering the icon is being scaled, and probably not too difficult to add. (I remember previous iterations of the patch where it wasn't)

I'm sure there's things you can do design-wise to make it look nicer, but I would be happy enough with the basics. More toggles can always be added later.

**Comment by Kevinjil** [comment]
> If you need a screenshot, this is what my tab bar looks like right now (shame about ZapZap not setting an icon): <img alt="Image" width="2000" height="45" src="https://private-user-images.githubusercontent.com/1339276/498801033-274686ea-ef5c-437e-a77c-ab7b1d803636.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTk5Mjc5MTEsIm5iZiI6MTc1OTkyNzYxMSwicGF0aCI6Ii8xMzM5Mjc2LzQ5ODgwMTAzMy0yNzQ2ODZlYS1lZjVjLTQzN2UtYTc3Yy1hYjdiMWQ4MDM2MzYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTAwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTEwMDhUMTI0NjUxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9N2JlN2U1YTJiM2RiMGVjNmI2M2ExMzJhNjYzYzRjNjUxYTM1YWEwYzE2ZmYwNGYzNGIwNjBlNjc5MTFkZTQxNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.De_-zX3sdc1XC219FFnIHKc67GiFIF4tpCtg5NlRJOw"> And I have some more screenshots with them [here](https://github.com/mid-kid/RocketISO/blob/master/docs/RELEASE-2025.md#screenshots).

This is very similar to my old i3 tab bar. I've been wishing to get these icons back since I switched.

> Besides, although it's just a visual nitpick, the icons displayed in your screenshot are too big for the titlebar. These minor details are something that also need to be considered: What's the ratio of icon vs text size? Would this feature have a way to specify icon size? Would the icons get bigger if the titlebar gets taller? What about the left and right margins, especially the space between the icons and the text?

I actually like that the icons fill all available vertical height, even outside the text padding. It makes them just this slightly bigger without losing screen space.

**Comment by darkralts** [comment]
```
warn: wayland.c:1734: compositor does not implement the XDG toplevel icon protocol
```
yeah not yet.

also just noting that as sway's main goal is also a drop in replacement for i3, as already stated this has been in mainstream i3 for years, and moving from i3 to sway i feel it would match its goal. and that people actually have a use outside of just aestetics for this, i feel its a worthy add.

**Comment by darkralts** [comment]
of course yes it would be optional and disabled by default if implemented.
i think we all feel the same on that note. plus thats how it is in i3

**Comment by Alex-ABPerson** [comment]
Just my two cents on the sub-discussion in here about whether it should be a thing to begin with.

I might not be understanding something in here, and that's completely my bad if so, but with the technical discussion aside, I'm somewhat amazed we seem to be even debating whether _window icons_ (the thing that basically every standard desktop environment/compositor in the world has) should be a thing. Icons on program's toplevels have been in UI for a long time, and it's a very standard thing almost any regular user would expect. I get that Sway isn't necessarily targeted at people like that, but I'd have to imagine the quantity of users who would complain about the enhancement would be _far_ lower than those who don't. It is literally free clarity/richness on your tab-bar out of the box. I'd even argue the kinds of people who use a compositor like this are those who would just know or subconsciously memorise icons and benefit from that to get a faster scan of their tabbars.

At the end of the day, my priority when I use Sway isn't to make my machine look pretty, it's to just have a genuinely powerful way of interacting and moving through my applications. The icons can be as inconsistent and crap as they want to be between programs, I don't get why that's really Sway's problem, but they still hold meaning anyway. And if making Sway pretty _is_ your priority, then I feel like you'll already be customizing a bunch that disabling them is just a few steps away.

For me the utility on a machine is the programs you use on it, and if a program you're using _asks_ for an icon to be shown on its titlebar, and it's a reasonable request for the compositor to fulfil (i.e. it fits into the environment the compositor provides, which it obviously does in this case, we all know where it 'would' go) surely the "right thing" _is for the compositor to honour that wish_, until the user asks it not to. I don't get why we'd consciously choose to lossily present the programs the user is using without an environmental reason to.

From a technical perspective: I _completely_ agree that when it wasn't a Wayland protocol it was a very unreasonable expectation, because that _is_ frankly just not in Sway's scope, and has that myriad of issues discussed.

And ALL of that being said..... I _do_ appreciate why there's the concern for attention to detail in getting it right if it is done, and not wanting to implement it until visual design has been nailed down. Hopefully we can agree on something there at some point, because I do think there's value in it!

_(Also I realize after writing this I talk above in the hypothetical of it being **on** by default, but I don't really feel all that strongly either way and it doesn't really change the core things, so since people said that they wanted off by default ofc that's probably the thing to do)_

**Comment by tkna91** [comment]
In Windows Explorer, it was standard for the application icon to appear in the top-left corner of the window, but I recall that macOS Finder windows did not have icons, and I don’t think it was possible to configure them (at least not using the standard methods).
I don’t feel that having an application icon in the window is particularly universal.

Especially when considering the wide range of potential applications for Wayland compositors and windows—such as Launching nested sway within sway, kiosk mode, window-only transmission via Waypipe, and other possibilities that haven’t yet been conceived—I find the principle of “keeping things as simple as possible” to be highly compelling. In my naive view, I suspect there are many factors to consider when deciding how to implement and define the display of application icons (via the sway, wloots, or Wayland protocols) within this context.

While users have a wide range of UI preferences, I believe Sway is equipped with flexible yet rigorous mechanisms that allow it to accommodate a broad spectrum of requirements, particularly those that prioritize low resource usage, lightweight performance, and speed.
There are likely many scenarios where "UI policies and use cases that do not require the application icon to be displayed in the title bar" apply. In such situations, you would naturally want to avoid consuming resources on features that are not needed.

That said, I do use the following method to display the icon. Since I can identify at a glance which application the window belongs to, I find it very convenient, at least for me personally.

https://github.com/iguanajuice/sway-font-awesome

On the other hand, the current version of labwc seems to display the application icon in the title bar quite well. I wonder how they’re doing it.

https://github.com/user-attachments/assets/9a45a199-e1bb-4f5a-ba21-f755358c1b98

**Comment by Alex-ABPerson** [comment]
Pure titlebars perhaps, but remember that Sway's titlebars double as a switching UI. And when you look at that, icons have always been a thing. Both the Windows taskbar and the Mac OS dock have them, even relying on them entirely now by default. I know "what everyone else is doing" doesn't always mean it's the best or right, I just think it's a thing to add to the consideration.

(I am never replying via email again good lord)

**Comment by David0tt** [comment]
I have implemented the functionality to show application icons in titlebars using a hacky solution that generates an icon font (like an emoji font) from all applications on the system on the fly and then runs a python daemon that automatically renames application titles to use the glyphs from this icon font, thereby showing application icons in the titlebar.

Originally I found this issue when searching for a way to put the icons of running programs into my workspace names, so I could easily identify my workspaces. I had implemented this and realized that one could use a similar solution to get icons into the titlebar so I added this functionality.

You can find my code that does both here: https://github.com/David0tt/workspace-icon-daemon

Here is a demo of how this looks:

https://github.com/user-attachments/assets/6c3e8dd3-2d3d-4fa9-b085-e7506e2c68ca

This is similar to https://gist.github.com/dmelliot/437924ff581f3f1edd59f44833be6cc6, https://github.com/iguanajuice/sway-font-awesome and https://github.com/swaywm/sway/issues/4882#issuecomment-611464474, however instead of using a preexisting font like Nerdfont or Fontawesome, a font with the true application icons is created on the fly. 

I hope I am not sidetracking this discussion. I think using such an add-in is a nice solution, and I fully understand why sway has good reasons to not implement this natively. In any case, I think this at least provides a demo of how this feature could potentially look and a workable solution for the people who really want this feature.

Some comments:
- Personally I don't think this turns into pixelmush (https://github.com/swaywm/sway/issues/4882#issuecomment-3380680791). Imho the resolution of the icons is quite good. The icon font only uses glyphs of 109x109, however on my 4k monitor it looks very sharp to me. Also, most programs I am using daily have very readable icons. 
- Personally I find icons much more readable to identify workspaces and running programs, so I see good arguments sway to natively implement icon support. If I did not have icons in my workspaces I would not be able to use sway as efficiently. My personal preference is to not use titlebars at all, distribute the programs over the workspaces and identify the appropriate workspace to go to from its icons. However with a layered titlebar layout I would also rely a lot on the icons.

**Comment by Alex-ABPerson** [comment]
Yeah, that looks great imo, honestly. Maybe it's just my tastes but I literally think the desktop looks _more vibrant_ visually with the icons than without (and I'm not even prioritising asethetics).

I've been meaning to for a while now and never found the time, but at some point I do plan to make a fork with support for xdg-toplevel-icon-v1 added (since afaik it's already in WLR and everything), just as an experiment to see how it feels.

**Comment by Konpaku-LVX** [comment]
> Yeah, that looks great imo, honestly. Maybe it's just my tastes but I literally think the desktop looks _more vibrant_ visually with the icons than without (and I'm not even prioritising asethetics).
> 
> I've been meaning to for a while now and never found the time, but at some point I do plan to make a fork with support for xdg-toplevel-icon-v1 added (since afaik it's already in WLR and everything), just as an experiment to see how it feels.

awesome let me know how that goes
